apm-commons/apm-datacarrier/src/main/java/org/apache/skywalking/apm/commons/datacarrier/consumer/ConsumerThread.java

```diff
@@ -36,7 +37,7 @@
         super(threadName);
         this.consumer = consumer;
         running = false;
-        dataSources = new LinkedList<DataSource>();
+        dataSources = new ArrayList<DataSource>(1);
```

Comment:
ArrayList size 1? Why?

Reply:
Because I think the default value(10) is large. It will only grow at initialization. Could you give some suggestion?

Reply:
This is why we use LinkedList. From several Google articles, we should always use ArrayList(no LinkedList)

But in this case, we can't be sure which is the right size for ArrayList, and how to be sure ArrayList is better in all scenario.

I used to think about use ArrayList in TracingContext, another core using LinkedList today. But I find out it is hard to make a reasonable decision.

Reply:
In DataCarrier Consumer, there are two
1. ConsumerThread(agent used today)
1. MultipleChannelsConsumer(OAP used today)

In the queue, it is much easier than TracingContext, you need to tests the performance in different initial sizes and different max sizes.

Such as in `MultipleChannelsConsumer`
1. Init=10, Expected list size=40000. In 16CPU VM, 32 Threads deal with 100 queues,  meaning max list size = 10000 * 4(queue size). Read `MultipleChannelsConsumer`
1. Init=1, Max data size=40000
1. Init = 40000, Max = 40000
1. Init = 8000, Max = 40000

Another optimization option is, hold the consumer list, just set the element to `null`, to avoid the size extension, but cost more GC potentially.

Reply:
Anyway, if you want to do this, I think we should work on a test blog first, work on different scenarios, then back to this PR. If this makes sense to you.

cc @zhaoyuguang @kezhenxu94 @ascrutae @JaredTan95   in case you are interested in this core optimization.

Reply:
Thanks for your comments. I will think more and provide more completely benchmark cases later.
Welcome to join.

Reply:
Feel free to create a new repo for set up benchmarks, if that is easier to create  benchmarks before change codes.

Reply:
I had tested the performance of the ArrayList and LinkedList. But the report was lost. I just remember that LinkedList is faster than ArrayList if the size is below 5w.
A very important thing you must know is Datacarrier sends it's data to the consumer 20ms. So, if the queue size more than 5w, that means the TPS is 5w * (1000ms / 20ms) = 250w. This is a not reachable TPS.

@dmsolr